src/circuit/gadget/ecc/chip.rs

```diff
@@ -0,0 +1,473 @@
+    pub q_mul_fixed: Selector,
+    /// Fixed-base signed short scalar multiplication
+    pub q_mul_fixed_short: Selector,
+    /// Witness point
+    pub q_point: Selector,
+    /// Witness full-width scalar for fixed-base scalar mul
+    pub q_scalar_fixed: Selector,
+    /// Witness signed short scalar for full-width fixed-base scalar mul
+    pub q_scalar_fixed_short: Selector,
+    /// Permutation
+    pub perm: Permutation,
+}
+
+/// A chip implementing EccInstructions
+#[derive(Clone, Debug)]
+pub struct EccChip<C: CurveAffine> {
+    id: u64,
+    config: EccConfig,
+    loaded: EccLoaded<C>,
+}
+
+impl<C: CurveAffine> PartialEq for EccChip<C> {
+    fn eq(&self, other: &Self) -> bool {
+        self.id == other.id
+    }
+}
+
+impl<C: CurveAffine> Eq for EccChip<C> {}
+
+impl<C: CurveAffine> EccLoaded<C> {
+    fn get(&self, point: OrchardFixedBases<C>) -> OrchardFixedBase<C> {
+        match point {
+            OrchardFixedBases::CommitIvkR(_) => self.commit_ivk_r.clone(),
+            OrchardFixedBases::NoteCommitR(_) => self.note_commit_r.clone(),
+            OrchardFixedBases::NullifierK(_) => self.nullifier_k.clone(),
+            OrchardFixedBases::ValueCommitR(_) => self.value_commit_r.clone(),
+        }
+    }
+
+    fn get_short(&self, _point: OrchardFixedBasesShort<C>) -> OrchardFixedBaseShort<C> {
+        self.value_commit_v.clone()
+    }
+}
+
+impl<C: CurveAffine> Chip<C::Base> for EccChip<C> {
+    type Config = EccConfig;
+    type Loaded = EccLoaded<C>;
+
+    fn config(&self) -> &Self::Config {
+        &self.config
+    }
+
+    fn loaded(&self) -> &Self::Loaded {
+        &self.loaded
+    }
+}
+
+impl<C: CurveAffine> EccChip<C> {
+    pub fn construct(
+        config: <Self as Chip<C::Base>>::Config,
+        loaded: <Self as Chip<C::Base>>::Loaded,
+    ) -> Self {
+        Self {
+            id: rand::random::<u64>(),
+            config,
+            loaded,
+        }
+    }
+
+    #[allow(non_snake_case)]
+    pub fn configure(
+        meta: &mut ConstraintSystem<C::Base>,
+        bits: Column<Advice>,
+        P: (Column<Advice>, Column<Advice>),
+        lambda: (Column<Advice>, Column<Advice>),
+        extras: [Column<Advice>; 5],
+    ) -> <Self as Chip<C::Base>>::Config {
+        let config = EccConfig {
+            bits,
+            P,
+            lambda,
+            extras,
+            lagrange_coeffs: [
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+                meta.fixed_column(),
+            ],
+            fixed_z: meta.fixed_column(),
+            #[cfg(test)]
+            q_double: meta.selector(),
+            q_add_incomplete: meta.selector(),
+            q_add: meta.selector(),
+            q_mul_hi: meta.selector(),
+            q_mul_lo: meta.selector(),
```

Comment:
Let's explore whether or not the Rust compiler guarantees that it will execute these operations in a particular and deterministic order. These calls happen within a single struct initialization expression, whereas in other code like this we usually separate them out into separate expressions which are guaranteed to run in a deterministic order by the language. If, for example, a new version of the Rust compiler is allowed to start initializing deeply-nested expressions before calling others, it could change the order of this and still compile successfully, breaking the mapping between the columns and thus failing to correctly verify and create proofs created by other users.

(I don't know if this guarantee is made by the compiler right now, if it is then we're good.)

Reply:
I couldn't find official documentation on whether this is guaranteed now / in future. There's a [StackOverflow thread](https://stackoverflow.com/questions/61984925/is-the-order-in-which-struct-fields-are-initialized-guaranteed-in-rust) which suggests that the order of field initialisation follows the order in which they're written in the initialiser. But failing any official documentation, I'm just going to separate this nested initialiser into single expressions.